WRQ-9365: Added VirtualGridList sample to load data incrementally for infinite scrolling

mmyelyn opened the pull request:
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
I have a request on how virtualgridlist should be implemented if data is continuously loaded.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
I used VirtualGridList's onScrollStop handler to load data when the scroll reaches the bottom.
Assuming that data has been fetched, I implemented it to attach mock data.


### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

### Links
[//]: # (Related issues, references)
WRQ-9365

### Comments
Enact-DCO-1.0-Signed-off-by: Hyelyn Kim ([myelyn.kim@lge.com](mailto:sc.baek@lge.com))

mmyelyn committed:
Added a sample with an infinite scroll to the virtual grid list

- Since it is a guide to infinite scroll, I added it as a pattern sample
- The data fetching part assumes success and appends the data.

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Updated README

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Modified the code location appropriately

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Cleaned up the code

- Cleaned up README, style and remove package-lock file

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Cleaned up the mainView

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

juwonjeong commented on sandstone/pattern-virtualgridlist-infinite-scroll/resources/ilibmanifest.json:
```diff
@@ -0,0 +1,3 @@
+{
+	"files": []
+}
```
Please add a new line

juwonjeong commented on sandstone/pattern-virtualgridlist-infinite-scroll/src/App/package.json:
```diff
@@ -0,0 +1,3 @@
+{
+	"main": "App.js"
+}
```
please add a new line

juwonjeong commented on sandstone/pattern-virtualgridlist-infinite-scroll/src/views/MainView.js:
```diff
@@ -0,0 +1,89 @@
+import {VirtualGridList} from '@enact/sandstone/VirtualList';
+import {useCallback, useState, useEffect} from 'react';
+
+const defaultSize = 10;
```
I think defaultSize needs to be fixed because in FHD or 4k dimensions, scroll does not appear.

mmyelyn replied:
Fixed defaultSize to 30 to cover resolutions for all screenTypes. The added data size was kept at 10 for ease of testing.

juwonjeong commented on sandstone/pattern-virtualgridlist-infinite-scroll/src/views/MainView.js:
```diff
@@ -0,0 +1,89 @@
+			itemSize={{minHeight: ri.scale(570), minWidth: ri.scale(688)}}
+			onScrollStop={handleScrollStop}
+		/>
+		</Panel>
```
Please fix the indentation

juwonjeong commented:
Please check the enact lint --strict in this sample

juwonjeong commented on sandstone/pattern-virtualgridlist-infinite-scroll/README.md:
```diff
@@ -0,0 +1,18 @@
+- `sandstone/Panels/Panel`
+- `sandstone/VirtualList/VirtualGridList`
+
+In this app, load data when scrolling reaches the bottom of the virtualgridlist. Use your API server to data fetch to load the data.
```
```suggestion
In this app, data is loaded when scrolling reaches the bottom of the virtualgridlist. Use your API server to fetch data to load.
```

juwonjeong commented on sandstone/pattern-virtualgridlist-infinite-scroll/README.md:
```diff
@@ -0,0 +1,18 @@
+- `sandstone/VirtualList/VirtualGridList`
+
+In this app, load data when scrolling reaches the bottom of the virtualgridlist. Use your API server to data fetch to load the data.
+Performance may decrease when using the onScroll handler, so use the onScrollStop handler like in the sample.
```
```suggestion
Performance may decrease when using the `onScroll` handler, so use the `onScrollStop` handler like in the sample.
```

juwonjeong commented:
Could you please add this pattern sample link to all-samples as other samples?
https://github.com/enactjs/samples/blob/master/sandstone/all-samples/src/App/App.js

mmyelyn committed:
Merge branch 'develop' into feature/WRQ-9365

mmyelyn committed:
Applied reviews

-fixed indentations and lint
-modified default size for variable screen types

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Added infinite scroll sample to all-samples

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Fixed lint warning

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

juwonjeong committed:
Update dependencies

Enact-DCO-1.0-Signed-off-by: Juwon Jeong (juwon.jeong@lge.com)

juwonjeong reviewed (approved):
LGTM

seunghoh commented on sandstone/pattern-virtualgridlist-infinite-scroll/src/views/MainView.js:
```diff
@@ -0,0 +1,89 @@
+import {useCallback, useState, useEffect} from 'react';
+
+const defaultSize = 30;
+const addedListSize = 10;
```
Let's use defaultSize to 50 and addedListSIze to 30 to see scroll position changing more obviously.

seunghoh commented on sandstone/pattern-virtualgridlist-infinite-scroll/src/reportWebVitals.js:
```diff
@@ -0,0 +1,13 @@
+	}
+};
+
+export default reportWebVitals;
```
We don't use WebVitals on samples.

seunghoh commented on sandstone/pattern-virtualgridlist-infinite-scroll/src/index.js:
```diff
@@ -0,0 +1,24 @@
+import {createRoot, hydrateRoot} from 'react-dom/client';
+
+import App from './App';
+import reportWebVitals from './reportWebVitals';
```
We don't use WebVitals on this samples repo.

seunghoh commented on sandstone/all-samples/src/App/App.js:
```diff
@@ -18,6 +18,7 @@ import {AppBase as PatternSinglePanel} from '../../../pattern-single-panel/src/A
 import {appElementBase as PatternSinglePanelRedux} from '../../../pattern-single-panel-redux/src/main';
 import {AppBase as PatternVideoPlayer} from '../../../pattern-video-player/src/App/App';
 import {appElementBase as PatternVirtualgridlistApi} from '../../../pattern-virtualgridlist-api/src/main';
+import {App as PatternVirtualgridlistInfiniteScroll} from '../../../pattern-virtualgridlist-infinite-scroll/src/App/App';
```
Let's use `pattern-virtualgridlist-incremental-load`. `infinite-scroll` is not the main topic on this sample. 
Please change the folder name as well.

mmyelyn replied:
That's a good point. It's changed. I also modified the sample's title.

seunghoh commented on sandstone/pattern-virtualgridlist-infinite-scroll/README.md:
```diff
@@ -0,0 +1,18 @@
+## VirtualGridList With Infinite Scroll Pattern
+
+A sample Enact application that shows off how to load items with infinite scroll in VirtualGridList.
+
+Run `npm install` then `npm run serve` to have the app running on [http://localhost:8080](http://localhost:8080), where you can view it in your browser.
+
+#### Enact Components Used
+- `sandstone/ImageItem`
+- `sandstone/Panels/Header`
+- `sandstone/Panels/Panel`
+- `sandstone/VirtualList/VirtualGridList`
+
+In this app, data is loaded when scrolling reaches the bottom of the virtualgridlist. Use your API server to fetch data to load.
+Performance may decrease when using the `onScroll` handler, so use the `onScrollStop` handler like in the sample.
+
+---
+
+This project was bootstrapped with the Enact [cli](https://github.com/enactjs/cli).
```
```suggestion
## Incremental Data Loading in VirtualGridList Pattern

A sample Enact application that demonstrates how to load items incrementally in a VirtualGridList.

Run `npm install` then `npm run serve` to have the app running on [http://localhost:8080](http://localhost:8080), where you can view it in your browser.

#### Enact Components Used
- `sandstone/ImageItem`
- `sandstone/Panels/Header`
- `sandstone/Panels/Panel`
- `sandstone/VirtualList/VirtualGridList`

In this app, data is loaded as you scroll to the bottom of the VirtualGridList. Use your API server to fetch data to load.
Performance may decrease when using the `onScroll` handler, so use the `onScrollStop` handler as shown in the sample.

---

This project was bootstrapped with the Enact [cli](https://github.com/enactjs/cli).
```

seunghoh reviewed (changes requested):
Let's change the sample name.

mmyelyn committed:
Applied the reviews

- Renamed the folder and file name appropriately, removed unnecessary files

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Removed incorrectly added files

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn commented:
> Let's change the sample name.

The sample name and other reviews have all been applied.

mmyelyn committed:
Modified package name

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)

mmyelyn committed:
Fixed lint warning

Enact-DCO-1.0-Signed-off-by: Hyelyn Kim (myelyn.kim@lge.com)
